### ISSUE TITLE
Module failing to instantiate with unknown provider error due to minification

### ISSUE THREAD

**Issue by palenshus** [vsternbach/angular-ts-decorators#45]
Hi! Angular-ts-decorators was working pretty well for me, but I just tried finally compiling with webpack's -p (minify) option, and now I'm getting an error:

`Uncaught Error: [$injector:modulerr] Failed to instantiate module AppModule due to:
Error: [$injector:unpr] Unknown provider: e`

I had a .config and a .run method on my module, I ended up getting rid of the .run method to simplify things, and here's all I have left. From reading previous issues, it looks like this library should be automatically providing the $inject property for me, but I get it with or without my added property.

```
export class AppModule {
    public static $inject = ["$urlRouterProvider", "$stateProvider"];

    public static config($urlRouterProvider: ng.ui.IUrlRouterProvider, $stateProvider: ng.ui.IStateProvider) {
        AppModule.provideStates(routes, $stateProvider);
        $urlRouterProvider.otherwise("/");
    }

    private static provideStates(states: IComponentState[], $stateProvider: ng.ui.IStateProvider) {
```

Here's (most) of the compiled .js code:

```
var m = function () {
    function e() {}
    return t = e,
    e.config = function (e, n) {
        t.provideStates(f, n),
        e.otherwise("/")
    },
...
    e.$inject = ["$urlRouterProvider", "$stateProvider"],
    e = t = i([o.NgModule({
                    id : "AppModule",
                    imports : ["legacyModule", "ui.router", "ui.bootstrap", "kendo.directives"],
                    declarations : [a.RootComponent, s.NavbarComponent, l.AppsComponent, c.AppDetailsComponent, u.AppVersionDetailsComponent, d.ProgressBarComponent],
                    providers : [h.ChartService, p.ApiService]
                })], e);
    var t
}
();
t.AppModule = m,
o.platformBrowserDynamic().bootstrapModule(m)
```

Any idea what the issue could be? I'm near my wit's end :) Thanks!

**Comment by vsternbach** [comment]
Where do you use this module and how? Can you show this code here?
I suppose the problem is in bootstrapModule, you need to pass module name as the argument
(<NgModuleDecorated>AppModule).module.name or as a string 'AppModule'

**Comment by dwadezp** [comment]
we have the same issue :(

**Comment by vsternbach** [comment]
@dwadezp can you provide more details on how and where you consume your module?

**Comment by dwadezp** [comment]
import { NgModule } from 'angular-ts-decorators';
import { SomeComponent } from './some.component';
import { SomeService } from './some.service';

@NgModule({
    id: 'SomeModule',
    declarations: [
        SomeComponent
    ],
    imports: [],
    providers: [
        {
            provide: 'someService', 
            useClass: SomeService
        }
    ]
})
export class SomeModule {}

Works perfectly with     new webpack.optimize.UglifyJsPlugin({
        mangle: false
    })

but with 
    new webpack.optimize.UglifyJsPlugin({
        mangle: true
    })

it crashes

:(

**Comment by dwadezp** [comment]
So, so for  most cases it was our fault, sorry. 
Could you please suggest hot to annotate such things?
            provide: 'SomeService', 
            useFactory:(
                $q,
                $log
            )=> new SomeService(
                $q,
                $log
            )

? We have some problems with minification in such cases

**Comment by vsternbach** [comment]
Ok
Regarding factories, there's propety deps for registering factories, that is not implemented for some reason, I'll add this implementation, with it you would be able to register it like this:
```
{
  provide: 'SomeService',
  useFactory: ($q, $log) => new SomeService($q, $log),
  deps: ['$q', '$log']
}
```
For now, assuming you are using ngAnnotate you can do like this:
```
/*@ngInject*/
export function someFunc($q, $log) {
  return new SomeService($q, $log);
}
```
and then register it like this: {provide: 'SomeService', useFactory: someFunc}

**Comment by dwadezp** [comment]
Cool thank you 👍

**Comment by vsternbach** [comment]
fixed in f51eea9cb3d45a0a4b8b3ab0dc95a95d036873dc
